### PR TITLE
Initialize sources as stream and fixes abort

### PR DESCRIPTION
**Pull request by Milly**
- Source initialization and items gathering asynchronously, and reduces the time it takes to show the first results.
- Abort the asynchronous process using the signal at the start, which allows for correct cancellation.

This PR has a lot of changes. Therefore, I would like to keep it as a draft until it is confirmed that there are no problems with execution.

### CODE REVIEW

File: denops/ddu/ddu.ts [hunk 1]
```diff
@@ -488,66 +662,93 @@ export class Ddu {
...
       // Already redrawing, so adding to schedule
+      const { restoreItemState, signal } = this.#scheduledRedrawOptions ?? {};
       this.#scheduledRedrawOptions = {
         // Override with true
-        restoreItemState: opts?.restoreItemState ||
-          this.#scheduledRedrawOptions?.restoreItemState,
+        restoreItemState: opts.restoreItemState || restoreItemState,
+        // Merge all signals
+        signal: signal && opts.signal && signal !== opts.signal
+          ? chainSignal(new AbortController(), signal, opts.signal).signal
```
Comment (Shougo) [inline denops/ddu/ddu.ts:729]: Please use `AbortSignal.any()` instead.

File: denops/ddu/ddu.ts [hunk 2]
```diff
@@ -247,36 +370,25 @@ export class Ddu {
...
 
-      await initSource(
-        denops,
-        source,
-        sourceOptions,
-        sourceParams,
-        this.#loader,
-      );
-    }
+    // Wait initialized all sources. Source onInit() must be called before UI
```
Comment (Shougo) [inline denops/ddu/ddu.ts:384]:
```suggestion
    // Wait until initialized all sources. Source onInit() must be called before UI.
```

File: denops/ddu/ddu.ts [hunk 3]
```diff
@@ -422,6 +549,53 @@ export class Ddu {
     }
   }
 
+  async #refreshItems(denops: Denops, state: GatherState): Promise<void> {
+    const { sourceInfo: { sourceOptions }, itemsStream } = state;
+
+    await this.#callOnRefreshItemsHooks(denops, sourceOptions);
+
+    // Get path option, or current directory instead if it is empty
```
Comment (Shougo) [inline denops/ddu/ddu.ts:557]:
```suggestion
    // Get path option, or current directory instead if it is empty.
```

File: denops/ddu/ddu.ts [hunk 4]
```diff
@@ -422,6 +549,53 @@ export class Ddu {
...
+    for await (const newItems of itemsStream) {
+      if (!equal(path, this.#context.path)) {
+        if (this.#context.path.length > 0) {
+          this.#context.pathHistories.push(this.#context.path);
+        }
+        this.#context.path = path;
+      }
+
+      if (!this.#options.sync && newItems.length > 0) {
+        // Do not await inside loop
```
Comment (Shougo) [inline denops/ddu/ddu.ts:571]:
```suggestion
        // Do not await inside loop.
```

File: denops/ddu/ddu.ts [hunk 5]
```diff
@@ -1213,76 +1452,74 @@ export class Ddu {
...
 
   async expandItem(
     denops: Denops,
     parent: DduItem,
     options: {
-      // Expand recursively to the maxLevel
-      maxLevel: number;
+      /**
+       * If specified, expand recursively to find it path.
+       * Otherwise, Expand recursively to the maxLevel
```
Comment (Shougo) [inline denops/ddu/ddu.ts:1484]:
```suggestion
       * Otherwise, Expand recursively to the maxLevel.
```

**Comment by Shougo** [comment]
LGTM.  It is still WIP?

**Comment by kyoh86** [comment]
It does not work well with ddu-source-lsp ("lspDefinitions" source).

Minimal config:
```
set runtimepath+=~/path/to/denops.vim
set runtimepath+=~/path/to/ddu.vim
set runtimepath+=~/path/to/ddu-ui-ff
set runtimepath+=~/path/to/ddu-source-lsp
set runtimepath+=~/path/to/nvim-lspconfig

" TODO: Setup nvim-lsp for any language server (e.g. gopls)
" lua require('lspconfig').gopls.setup({})

command DduLspDefinition call ddu#start({
\   "ui": "ff",
\   "sync": 1,
\   "sources": [
\     { "name": "lsp_definition", "params": { "method": "textDocument/definition" } },
\     { "name": "lsp_definition", "params": { "method": "textDocument/typeDefinition" } },
\     { "name": "lsp_definition", "params": { "method": "textDocument/declaration" } },
\     { "name": "lsp_definition", "params": { "method": "textDocument/implementation" } },
\   ],
\   "kindOptions": {
\     "lsp": {
\       "defaultAction": "open",
\     }
\   }
\ })
```

With commands to install plugins:

```console
$ git clone "https://github.com/Milly/ddu.vim" -b source-stream
$ git clone "https://github.com/vim-denops/denops.vim"
$ git clone "https://github.com/Shougo/ddu-ui-ff"
$ git clone "https://github.com/uga-rosa/ddu-source-lsp"
$ git clone "https://github.com/neovim/nvim-lspconfig"
```
And other requirements are any LSP, such as gopls, and source code in the corresponding language.

Calling `DduLspDefinition`, I could not get any items to be shown.
But I use the Shougo/ddu.vim instead, they are shown.

```console
$ git clone "https://github.com/Shougo/ddu.vim"
```

It may be an implementation issue with ddu-source-lsp, but I am just reporting the fact that it occurred.

**Comment by Shougo** [comment]
Reproduced.

**Review by Shougo — CHANGES_REQUESTED** [review]
`ddu-source-lsp` returns items, but it is not displayed.  It is bug.

**Comment by Shougo** [comment]
```diff
diff --git a/denops/ddu/ddu.ts b/denops/ddu/ddu.ts
index 46aa462..2842f2c 100644
--- a/denops/ddu/ddu.ts
+++ b/denops/ddu/ddu.ts
@@ -184,6 +184,8 @@ class GatherState<
     const appendStream = new TransformStream<DduItem[], DduItem[]>({
       transform: (newItems, controller) => {
         this.#items = this.#items.concat(newItems);
+        console.log("items=");
+        console.log(this.#items);
         controller.enqueue(newItems);
       },
       flush: () => {
@@ -697,6 +699,8 @@ export class Ddu {
 
         // Wait until the stream closes.
         await itemsStream.pipeTo(itemTransformer.writable);
+        console.log(source.name);
+        console.log("Wait until the stream closes.");
       } catch (e: unknown) {
         if (state.signal.aborted && e === state.signal.reason) {
           // Aborted by signal, so do nothing.
@@ -2240,6 +2244,7 @@ export class Ddu {
     if (!state || !source) {
       return [false, 0, []];
     }
+    console.log(state.items);
 
     // NOTE: Use deepcopy.  Because of filters may break original items.
     let items = structuredClone(state.items) as DduItem[];
[~/work/ddu.vim] source-stream|+1
% git reset --hard                                     
HEAD is now at b802c71 feat!: Deno 1.40+ is required.
[~/work/ddu.vim] source-stream|✔
% git switch main         
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
[~/work/ddu.vim] main|✔
% git switch source-stream
Switched to branch 'source-stream'
[~/work/ddu.vim] source-stream|✔
% git diff                
diff --git a/denops/ddu/ddu.ts b/denops/ddu/ddu.ts
index 46aa462..e884e13 100644
--- a/denops/ddu/ddu.ts
+++ b/denops/ddu/ddu.ts
@@ -76,9 +76,6 @@ import { defaultActionOptions } from "./base/action.ts";
 import { Loader } from "./loader.ts";
 import { errorException, treePath2Filename } from "./utils.ts";
 
-// deno-lint-ignore no-explicit-any
-type AnySource = BaseSource<any>;
-
 type ItemActions = {
   source: BaseSource<BaseSourceParams, unknown>;
   kind: BaseKind<BaseKindParams>;
@@ -184,6 +181,8 @@ class GatherState<
     const appendStream = new TransformStream<DduItem[], DduItem[]>({
       transform: (newItems, controller) => {
         this.#items = this.#items.concat(newItems);
+        console.log("this.#items=");
+        console.log(this.#items);
         controller.enqueue(newItems);
       },
       flush: () => {
@@ -235,7 +234,7 @@ class GatherState<
 
 export class Ddu {
   #loader: Loader;
-  readonly #gatherStates = new Map<AnySource, GatherState>();
+  readonly #gatherStates = new Map<number, GatherState>();
   #input = "";
   #context: Context = defaultContext();
   #options: DduOptions = defaultDduOptions();
@@ -520,7 +519,7 @@ export class Ddu {
           0,
           { signal },
         );
-        this.#gatherStates.set(source, state);
+        this.#gatherStates.set(sourceIndex, state);
 
         controller.enqueue(state);
       },
[~/work/ddu.vim] source-stream|+1
% git diff
diff --git a/denops/ddu/ddu.ts b/denops/ddu/ddu.ts
index 46aa462..e884e13 100644
--- a/denops/ddu/ddu.ts
+++ b/denops/ddu/ddu.ts
@@ -76,9 +76,6 @@ import { defaultActionOptions } from "./base/action.ts";
 import { Loader } from "./loader.ts";
 import { errorException, treePath2Filename } from "./utils.ts";
 
-// deno-lint-ignore no-explicit-any
-type AnySource = BaseSource<any>;
-
 type ItemActions = {
   source: BaseSource<BaseSourceParams, unknown>;
   kind: BaseKind<BaseKindParams>;
@@ -184,6 +181,8 @@ class GatherState<
     const appendStream = new TransformStream<DduItem[], DduItem[]>({
       transform: (newItems, controller) => {
         this.#items = this.#items.concat(newItems);
+        console.log("this.#items=");
+        console.log(this.#items);
         controller.enqueue(newItems);
       },
       flush: () => {
@@ -235,7 +234,7 @@ class GatherState<
 
 export class Ddu {
   #loader: Loader;
-  readonly #gatherStates = new Map<AnySource, GatherState>();
+  readonly #gatherStates = new Map<number, GatherState>();
   #input = "";
   #context: Context = defaultContext();
   #options: DduOptions = defaultDduOptions();
@@ -520,7 +519,7 @@ export class Ddu {
           0,
           { signal },
         );
-        this.#gatherStates.set(source, state);
+        this.#gatherStates.set(sourceIndex, state);
 
         controller.enqueue(state);
       },
...skipping...
         }),
     );
@@ -1508,7 +1512,7 @@ export class Ddu {
     if (source == null) {
       return;
     }
-    const state = this.#gatherStates.get(source);
+    const state = this.#gatherStates.get(sourceIndex);
     if (state == null) {
       return;
     }
@@ -1733,7 +1737,7 @@ export class Ddu {
         this.#options,
         this.#options.sources[index],
       );
-      const state = this.#gatherStates.get(source);
+      const state = this.#gatherStates.get(index);
       if (state == null) {
         continue;
       }
@@ -2226,6 +2230,7 @@ export class Ddu {
   async #filterItems(
     denops: Denops,
     userSource: UserSource,
+    sourceIndex: number,
     input: string,
   ): Promise<[boolean, number, DduItem[]]> {
     userSource = convertUserString(userSource);
@@ -2236,10 +2241,11 @@ export class Ddu {
       userSource,
     );
 
-    const state = this.#gatherStates.get(source!);
+    const state = this.#gatherStates.get(sourceIndex);
     if (!state || !source) {
       return [false, 0, []];
     }
+    console.log(state.items);
 
     // NOTE: Use deepcopy.  Because of filters may break original items.
     let items = structuredClone(state.items) as DduItem[];
[~/work/ddu.vim] source-stream|+1
% git diff
diff --git a/denops/ddu/ddu.ts b/denops/ddu/ddu.ts
index 46aa462..60e41ba 100644
--- a/denops/ddu/ddu.ts
+++ b/denops/ddu/ddu.ts
@@ -76,9 +76,6 @@ import { defaultActionOptions } from "./base/action.ts";
 import { Loader } from "./loader.ts";
 import { errorException, treePath2Filename } from "./utils.ts";
 
-// deno-lint-ignore no-explicit-any
-type AnySource = BaseSource<any>;
-
 type ItemActions = {
   source: BaseSource<BaseSourceParams, unknown>;
   kind: BaseKind<BaseKindParams>;
@@ -235,7 +232,7 @@ class GatherState<
 
 export class Ddu {
   #loader: Loader;
-  readonly #gatherStates = new Map<AnySource, GatherState>();
+  readonly #gatherStates = new Map<number, GatherState>();
   #input = "";
   #context: Context = defaultContext();
   #options: DduOptions = defaultDduOptions();
@@ -520,7 +517,7 @@ export class Ddu {
           0,
           { signal },
         );
-        this.#gatherStates.set(source, state);
+        this.#gatherStates.set(sourceIndex, state);
 
         controller.enqueue(state);
       },
@@ -796,6 +793,7 @@ export class Ddu {
           const [done, maxItems, items] = await this.#filterItems(
             denops,
             userSource,
+            index,
             this.#input,
           );
 
@@ -1039,8 +1037,8 @@ export class Ddu {
         .filter(([, state]) =>
           isRefreshTarget(state.sourceInfo.sourceIndex, refreshIndexes)
         )
-        .map(([source, state]) => {
-          this.#gatherStates.delete(source);
+        .map(([sourceIndex, state]) => {
+          this.#gatherStates.delete(sourceIndex);
           return state.waitDone;
         }),
     );
@@ -1508,7 +1506,7 @@ export class Ddu {
     if (source == null) {
       return;
     }
-    const state = this.#gatherStates.get(source);
+    const state = this.#gatherStates.get(sourceIndex);
     if (state == null) {
       return;
     }
@@ -1733,7 +1731,7 @@ export class Ddu {
         this.#options,
         this.#options.sources[index],
       );
-      const state = this.#gatherStates.get(source);
+      const state = this.#gatherStates.get(index);
       if (state == null) {
         continue;
       }
@@ -2226,6 +2224,7 @@ export class Ddu {
   async #filterItems(
     denops: Denops,
     userSource: UserSource,
+    sourceIndex: number,
     input: string,
   ): Promise<[boolean, number, DduItem[]]> {
     userSource = convertUserString(userSource);
@@ -2236,7 +2235,7 @@ export class Ddu {
       userSource,
     );
 
-    const state = this.#gatherStates.get(source!);
+    const state = this.#gatherStates.get(sourceIndex);
     if (!state || !source) {
       return [false, 0, []];
     }
[~/work/ddu.vim] source-stream|+1
% git diff
diff --git a/denops/ddu/ddu.ts b/denops/ddu/ddu.ts
index 46aa462..60e41ba 100644
--- a/denops/ddu/ddu.ts
+++ b/denops/ddu/ddu.ts
@@ -76,9 +76,6 @@ import { defaultActionOptions } from "./base/action.ts";
 import { Loader } from "./loader.ts";
 import { errorException, treePath2Filename } from "./utils.ts";
 
-// deno-lint-ignore no-explicit-any
-type AnySource = BaseSource<any>;
-
 type ItemActions = {
   source: BaseSource<BaseSourceParams, unknown>;
   kind: BaseKind<BaseKindParams>;
@@ -235,7 +232,7 @@ class GatherState<
 
 export class Ddu {
   #loader: Loader;
-  readonly #gatherStates = new Map<AnySource, GatherState>();
+  readonly #gatherStates = new Map<number, GatherState>();
   #input = "";
   #context: Context = defaultContext();
   #options: DduOptions = defaultDduOptions();
@@ -520,7 +517,7 @@ export class Ddu {
           0,
           { signal },
         );
-        this.#gatherStates.set(source, state);
+        this.#gatherStates.set(sourceIndex, state);
 
         controller.enqueue(state);
       },
@@ -796,6 +793,7 @@ export class Ddu {
           const [done, maxItems, items] = await this.#filterItems(
             denops,
             userSource,
+            index,
             this.#input,
           );
 
@@ -1039,8 +1037,8 @@ export class Ddu {
         .filter(([, state]) =>
           isRefreshTarget(state.sourceInfo.sourceIndex, refreshIndexes)
         )
-        .map(([source, state]) => {
-          this.#gatherStates.delete(source);
+        .map(([sourceIndex, state]) => {
+          this.#gatherStates.delete(sourceIndex);
           return state.waitDone;
         }),
     );
@@ -1508,7 +1506,7 @@ export class Ddu {
     if (source == null) {
       return;
     }
-    const state = this.#gatherStates.get(source);
+    const state = this.#gatherStates.get(sourceIndex);
     if (state == null) {
       return;
     }
@@ -1733,7 +1731,7 @@ export class Ddu {
         this.#options,
         this.#options.sources[index],
       );
-      const state = this.#gatherStates.get(source);
+      const state = this.#gatherStates.get(index);
       if (state == null) {
         continue;
       }
@@ -2226,6 +2224,7 @@ export class Ddu {
   async #filterItems(
     denops: Denops,
     userSource: UserSource,
+    sourceIndex: number,
     input: string,
   ): Promise<[boolean, number, DduItem[]]> {
     userSource = convertUserString(userSource);
@@ -2236,7 +2235,7 @@ export class Ddu {
       userSource,
     );
 
-    const state = this.#gatherStates.get(source!);
+    const state = this.#gatherStates.get(sourceIndex);
     if (!state || !source) {
       return [false, 0, []];
     }
```

Fixes the problem.

File: denops/ddu/ddu.ts [hunk 6]
```diff
@@ -106,25 +106,153 @@ type ItemActionInfo = {
...
+    if (this.itemsStream != null) {
+      await Array.fromAsync(this.itemsStream);
+    }
+  }
+}
+
 export class Ddu {
   #loader: Loader;
-  #gatherStates: Record<string, GatherState> = {};
+  readonly #gatherStates = new Map<AnySource, GatherState>();
```
Comment (Shougo) [inline denops/ddu/ddu.ts:238]: This is broken.  Please fix.

**Comment by Shougo** [comment]
Current problem:  `narrow` action is broken.  It seems the state is cached.

**Comment by kyoh86** [comment]
:memo::heart: I checked that the issue in my comment is resolved:

> It does not work well with ddu-source-lsp ("lspDefinitions" source) **in Neovim**.

**Comment by kuuote** [comment]
If refresh items with `ddu-option-sync`, got strange behavior.

# To Reproduce

1. source following config

```
set rtp+=/path/to/Milly/ddu.vim_source-stream
set rtp+=/path/to/Shougo/ddu-kind-file
set rtp+=/path/to/Shougo/ddu-source-file
set rtp+=/path/to/Shougo/ddu-ui-ff
set rtp+=/path/to/vim-denops/denops.vim

call ddu#custom#patch_global('ui', 'ff')
call ddu#custom#patch_global('sync', v:true)

nnoremap n <Cmd>call ddu#ui#do_action('itemAction', {'name': 'narrow', 'params': {'path': '..'}})<CR>

call ddu#start({'sources': [{'name': 'file'}]})
```

2. Press `n`

**Review by Shougo — APPROVED** [review]

**Comment by Shougo** [comment]
Merged.  Thanks.

**Pull request merged**